### PR TITLE
fix: retry interceptor now closes previous response

### PR DESCRIPTION
**Pull request by rjuliano**
*Issue #, if available:* #305

*Description of changes:*
The `RetryInterceptor` was not properly closing the response from a previous attempt before before re-trying the request by invoking `chain.proceed`.

This PR attempts to address the issue by closing the response for scenarios when the request will be retried. Added a couple of unit tests I used to reproduce the bug.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-android-sdk-appsync/src/test/java/com/amazonaws/mobileconnectors/appsync/retry/RetryInterceptorTest.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        mockWebServer.enqueue(new MockResponse().setResponseCode(200).setBody("{\"result\":\"all good\""));
+
+        Request request = new Request.Builder()
+            .url("http://localhost:8888")
+            .method("POST", RequestBody.create("{}", MediaType.get("application/json")))
+            .build();
+
+        final AtomicBoolean successful = new AtomicBoolean(false);
+        final CountDownLatch requestLatch = new CountDownLatch(1);
+        okHttpClient.newCall(request).enqueue(new Callback() {
```
[the remark this answers is not in the document]
Comment (rjuliano) [reply aws-android-sdk-appsync/src/test/java/com/amazonaws/mobileconnectors/appsync/retry/RetryInterceptorTest.java:75]: I started looking at that and unfortunately the Await utility is in the integration tests package. I was tinkering with moving it under the this package, but turned out to be more of a cluster than what it was worth. I'll leave it as is for now if that's OK

[the remark this answers is not in the document]
Comment (jamesonwilliams) [reply aws-android-sdk-appsync/src/test/java/com/amazonaws/mobileconnectors/appsync/retry/RetryInterceptorTest.java:75]: You can always just copy-paste it. It is unlikely to change so the code duplication is probably not worth optimizing for.

[the remark this answers is not in the document]
Comment (rjuliano) [reply aws-android-sdk-appsync/src/test/java/com/amazonaws/mobileconnectors/appsync/retry/RetryInterceptorTest.java]: Clipboard inheritance never fails me 🤣  The chain of callbacks made it a little 🤯 , but I think it looks a little cleaner.

2 line fix. 142 lines of unit tests. The CS gods are happy 😄

**Comment by rjuliano** [comment]
> Is this codebase able to make use of lambda expressions?

No...the test package has sourceCompatibility set to 1.8 but the rest of the codebase doesn't

**Review by jamesonwilliams — APPROVED** [review]

**Comment by jamesonwilliams** [comment]
> > Is this codebase able to make use of lambda expressions?
> 
> No...the test package has sourceCompatibility set to 1.8 but the rest of the codebase doesn't

Ah bummer!

**Pull request merged**